### PR TITLE
Simplify StartupProjectRegistrar and remove ProjectGuid property usage

### PR DESCRIPTION
**Pull request by davkean**
- Removed need to keep state if the project is debuggable or not, IVsStartupProjectsListService is designed to no op if the project has already been added or removed.
- Removed need to switch to UI thread. IVsStartupProjectsListService is free-threaded.
- Stop listening to project GUID changes - the project GUID rule is being removed to fix https://github.com/dotnet/project-system/issues/2940.
- Listen to all properties not just output type. If not using launch profiles, output type doesn't actually influence whether you are debuggable or not. It's based on the debugger properties such as executable, etc.

### CODE REVIEW

**Comment by davkean** [comment]
Need to fix tests.

**Comment by davkean** [comment]
There's a little more diff in this than I'd like because I rewrote it and then replaced the existing one. I'll reduce the diff there's concerns.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs [hunk 1]
```diff
@@ -2,150 +2,135 @@
...
     /// </summary>
-    internal class StartupProjectRegistrar : OnceInitializedOnceDisposed
+    internal class StartupProjectRegistrar : OnceInitializedOnceDisposedAsync
     {
+        private readonly IAsyncServiceProvider _serviceProvider;
         private readonly IProjectThreadingService _threadingService;
-        private readonly IActiveConfiguredProjectSubscriptionService _activeConfiguredProjectSubscriptionService;
+        private readonly IActiveConfiguredProjectSubscriptionService _projectSubscriptionService;
         private readonly ActiveConfiguredProject<DebuggerLaunchProviders> _launchProviders;
-        private readonly IVsService<IVsStartupProjectsListService> _startupProjectsListService;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: IVsService enforces the service be accessed on the UI thread - which is not want we want here. EIther I need to remove it and in lieu of IAsyncServiceProvider long term, or change its pattern to "async".

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs [hunk 2]
```diff
@@ -2,150 +2,135 @@
...
 
-        protected override void Initialize()
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
         {
-            var watchedEvaluationRules = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName);
-            var evaluationBlock = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(ConfigurationGeneralRuleBlock_ChangedAsync);
-            _evaluationSubscriptionLink = _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock
-                                            .LinkTo(target: evaluationBlock, ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates: true);
-        }
+            _projectGuid = await _projectGuidService.GetProjectGuidAsync()
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: We can assume the GUID never changes over the lifetime of a project one the factory has completed.

**Comment by davkean** [comment]
Fixed tests, ready for review.

**Comment by davkean** [comment]
Hit: https://github.com/dotnet/project-system/pull/3308.

@dotnet-bot retest this please

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs [hunk 3]
```diff
@@ -2,150 +2,138 @@
...
-                    await AddOrRemoveProjectFromStartupProjectListAsync(initialize: true).ConfigureAwait(false);
-                }
+            Assumes.Present(_startupProjectsListService);
 
-                return;
-            }
+            _subscription = _projectSubscriptionService.ProjectRuleSource.SourceBlock.LinkTo(
+                target: new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChangedAsync),
+                suppressVersionOnlyUpdates: true,
+                linkOptions: new DataflowLinkOptions() { PropagateCompletion = true });
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: Do you need to capture capabilities here?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: Good point. @lifengl thoughts? We read from a OrderPrecedenceImportCollection from the active configuration in the handler.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: I filed https://github.com/dotnet/project-system/issues/3322 to track this; I didn't change when we got initialized or when we walk the debuggers, so if the capabilities were wrong, they were always wrong. I'll do a pass over every dataflow usage and figure out if we need start a new context.

**Pull request merged**